### PR TITLE
shipit/api: authorization work

### PR DESCRIPTION
**Pull request by garbas**

### CODE REVIEW

File: lib/backend_common/backend_common/auth.py [hunk 1]
```diff
@@ -116,6 +116,35 @@ def has_permissions(self, permissions):
         return taskcluster.utils.scopeMatch(self.get_permissions(), permissions)
 
 
+class Auth0User(BaseUser):
+
+    type = 'auth0'
+
+    def __init__(self, token, userinfo):
+        assert 'email' in userinfo
```
Comment (rail) [inline lib/backend_common/backend_common/auth.py]: I may sound a bit paranoid, but can you avoid using `assert`. It always reminds me about [this issue](https://github.com/IdentityPython/pysaml2/issues/451) :)

**Comment by garbas** [comment]
@marco-c @La0 could you also pass a review for this PR. i will shortly push to staging so you can really test it.

the reason for involving you is that i refactor a bit the authentication part and while i'm sure i did it correctly i can never be 100%. that is why i would like additional review. also this PR might mess up some of your projects which makes it at least fair to bring it to your attention.

File: lib/backend_common/backend_common/auth.py [hunk 2]
```diff
@@ -184,7 +223,24 @@ def wrapper(*args, **kwargs):
...
-    def require_scopes(self, scopes):
+    def _require_permissions(self, permissions):
+        if not self._require_login():
+            return False
+
+        with flask.current_app.app_context():
+            if not flask_login.current_user.has_permissions(permissions):
+                user = flask_login.current_user.get_id()
+                user_permissions = flask_login.current_user.get_permissions()
+                diff = ' OR '.join([
```
Comment (La0) [inline lib/backend_common/backend_common/auth.py]: Isn't there a function in taskcluster-client that does this ?

File: src/shipit/api/shipit_api/api.py [hunk 3]
```diff
@@ -169,6 +145,12 @@ def schedule_phase(name, phase):
     except NoResultFound:
         abort(404)
 
+    # we must require scope which depends on product
+    required_permission = f'{SCOPE_PREFIX}/schedule_phase/{phase.release.product}/{phase.name}'
+    if not current_user.has_permissions(required_permission):
```
Comment (La0) [inline src/shipit/api/shipit_api/api.py]: :100:

File: lib/backend_common/backend_common/auth.py (hunk 2 above)
Comment (garbas) [reply lib/backend_common/backend_common/auth.py]: yes, but we might not be using taskcluser for authorization. like we do in `shipit/api`.

**Pull request merged**